libxfce4ui/xfce-gtk-extensions.c

```diff
@@ -49,6 +49,464 @@
+                                           callback_param, active, menu_to_append_item);
+    }
+  else
+    {
+      g_warning ("xfce_gtk_toggle_menu_item_new_from_action_entry: Unknown item_type");
+      return NULL;
+    }
+}
+
+
+
+/**
+ * xfce_gtk_tool_button_new_from_action_entry:
+ * @action_entry : Label to use for the #GtkTolButton
+ * @callback_param : optional callback parameter, or NULL.
+ * @toolbar_to_append_item : #GtkToolBar on which the item should be appended
+ *
+ * Method to create a toolbar button from the passed action entry
+ *
+ * Return value: (transfer full): A new #GtkTolButton
+ *
+ * Since: 4.16
+ **/
+GtkWidget*
+xfce_gtk_tool_button_new_from_action_entry (const XfceGtkActionEntry *action_entry,
+                                            GObject                  *callback_param,
+                                            GtkToolbar               *toolbar_to_append_item)
+{
+  GtkToolItem *tool_item;
+  GtkWidget   *image;
+
+  image = gtk_image_new_from_icon_name (action_entry->menu_item_icon_name, GTK_ICON_SIZE_LARGE_TOOLBAR);
+  tool_item = gtk_tool_button_new (image, action_entry->menu_item_label_text);
+  g_signal_connect_swapped (G_OBJECT (tool_item), "clicked", action_entry->callback, callback_param);
+  gtk_widget_set_tooltip_text (GTK_WIDGET (tool_item), action_entry->menu_item_tooltip_text);
+  gtk_widget_show (GTK_WIDGET (tool_item));
+  gtk_widget_show (image);
+  gtk_toolbar_insert (toolbar_to_append_item, tool_item, -1);
+  return GTK_WIDGET (tool_item);
+}
+
+
+
+/**
+ * xfce_gtk_menu_append_seperator:
+ * @menu : #GtkMenuShell on which the separator should be appended
+ *
+ * convenient method do add separators, used to prevent code duplication
+ *
+ * Since: 4.16
+ **/
+void
+xfce_gtk_menu_append_seperator (GtkMenuShell *menu)
+{
+  GtkWidget *item;
+
+  g_return_if_fail (GTK_IS_MENU_SHELL (menu));
+
+  item = gtk_separator_menu_item_new ();
+  gtk_menu_shell_append (menu, item);
+  gtk_widget_show (item);
+}
+
+
+
+/**
+ * xfce_gtk_accel_group_append:
+ * @accel_group : The #GtkAccelGroup to which the passed action_entries should be appended
+ * @action_entries : array of action_entries to be added
+ * @n_action_entries : size of the action_entries array
+ * @callback_data : callback data which should be passed if the accelerator is activated
+ *
+ * Connects the accel_pathes of the passed action_entries to the accel_group
+ * and adds the default key of each ActionEntry to the accel_map, if no key was defined for the related accel_path so far.
+ *
+ * Return value: (transfer none): The #GtkAccelGroup to which the passed action_entries were appended
+ *
+ * Since: 4.16
+ **/
+GtkAccelGroup*
+xfce_gtk_accel_group_append (GtkAccelGroup            *accel_group,
+                             const XfceGtkActionEntry *action_entries,
+                             guint                     n_action_entries,
+                             gpointer                  callback_data)
```

Comment:
What is callback_data used for?

Reply:
The calback itself is defined as en element of "XfceGtkActionEntry".
If you define e.g. "thunar_launcher_action_rename" as callback, you may want to pass and object of type ThunarLauncher*  as callback_data to that method. (e.g. to find out which files are selected, etc.)

Typically `ThunarLauncher.c` would define a List of XfceGtkActionEntry statically which would be added once to some action group. Like that all callbacks would require the same instance of ThunarLauncher* as callback_data.

Reply:
Oh, I see, you're basically copying `gtk_action_group_add_actions()` here.  Thanks for the explanation!
It is a little bit confusing that action entries are added to an accel group rather than an action group, though...